File: solutions/Golang/product_except_self_prefix_suffix.go
func productExceptSelf(nums []int) []int {
	n := len(nums)
	prefix := make([]int, n)
	suffix := make([]int, n)
	prefix[0] = 1
	suffix[n-1] = 1

	for i := 1; i < n; i++ {
		prefix[i] = prefix[i-1] * nums[i-1]
	}

	for i := n - 2; i >= 0; i-- {
		suffix[i] = suffix[i+1] * nums[i+1]
	}

	output := make([]int, n)
	for i := 0; i < n; i++ {
		output[i] = prefix[i] * suffix[i]
	}

	return output
}

File: solutions/Rust/product_except_self_prefix_suffix.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut prefix = vec![1i32; n];
        let mut suffix = vec![1i32; n];

        for i in 1..n {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        for i in (0..n - 1).rev() {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }

        (0..n).map(|i| prefix[i] * suffix[i]).collect()
    }
}
